### ISSUE TITLE
Make Validation page available to unauthenticated users

### ISSUE THREAD

**Issue by RakshindaAslam** [CDCgov/prime-reportstream#5517]
## Problem statement

When onboarding senders, validating files is a back-and-forth for O&O and the sender, and can take any varied amount of time. To mitigate this, **we can make the Validation service available to unauthenticated users**.

To effectively, and safely achieve this, prerequisites have been established:

- [ ] Work will be done to implement global "off switch" (see #7456 and #7390)
- [ ] Research will be done into where this fits into the onboarding experience
- [ ] Testing of the current validator with onboarding senders

## What you need to know

-  [Mural board for Validation Service Deep Dive with Options ](https://app.mural.co/t/primedatainput6410/m/primedatainput6410/1651702156098/60d77d324b3bf252c83d30f73f2cc201a59cc1eb?sender=u3fea607b46e84058d3224917)
 - Provide user-interface that allows file drag&drop or upload feature and display of response to the file
 - There is an expectation of reusing CSV Uploader components and features and not a full new paged
- [Figma Final Design (Eng-ready)](https://www.figma.com/file/6mwI5ac6rprACKDzDo4Ady/ReportStream-Workspace-%7C-2023?node-id=2183%3A21892&t=WLhbJIqQnfBOulOo-1)
- [Prototype of Final Design](https://www.figma.com/proto/6mwI5ac6rprACKDzDo4Ady/ReportStream-Workspace-%7C-2023?node-id=2183%3A21892&scaling=min-zoom&page-id=2173%3A22160&starting-point-node-id=2183%3A21892) 
- [Deprecated] Design details in https://app.zenhub.com/workspaces/experience-607d9d5e68b95200150fec37/issues/cdcgov/prime-reportstream/6365

## Acceptance criteria

- [ ] O&O can link onboarding senders to the validation service UI
- [ ] Onboarding senders with no Okta auth for login can access and use the validation service to test their CSV and HL7 files and view the response

## Screenshots
Old Feature Screenshot:
[reportstream.cdc.gov_file-handler_validate.png](https://images.zenhubusercontent.com/637d310dba428f3e3d55c9db/09dc27c8-660e-40b2-b8d0-82d2e12cefdd)

New Feature Screenshot:
[staging.reportstream.cdc.gov_file-handler_validate.png](https://images.zenhubusercontent.com/637d310dba428f3e3d55c9db/5691e8ce-5146-46b8-b8f7-6629c956bd5c)

**Comment by doug-s-nava** [comment]
a few notes at this point:
* RADx-MARS, as far as I know, is still out of scope for this service
* email notification should be a separate ticket, as that is all backend related work

**Comment by kevinhaube** [comment]
See associated Epic #6993 for progress on when we can release this feature to a wider audience

**Comment by kevinhaube** [comment]
@CnoteNava @chris-kuryak @HeatherRienks Updated the wording for the problem statement and AC to better reflect the narrow scope of _this_ ticket. Any additional work to achieve this goal is tracked in the ticket body with those check marks.

@fmccaf1 Please update the list items pertaining to research prerequisites with the proper epic numbers. Thank you!

**Comment by chris-kuryak** [comment]
@kevinhaube @fmccaf1 
This actually feels like it could have a big impact on a Key Result to reduce the amount of time the O+O team spends onboarding senders.
1) Is this issue fully defined yet, or should we have a quick meeting to iron out the details?
2) If it is fully defined, is this issue still blocked by something else, or could we begin work?

cc: @stephenkao

**Comment by stephenkao** [comment]
It looks like the validation page is locked to only admin users at the moment, and it shows a different format (CSV versus HL7) depending on the sender.  Also, at first glance, it looks like there'll be some API work to be done since it requires authentication -- or at least we'd want to clear it with the Platform team to make the change here: https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/src/main/kotlin/azure/ValidateFunction.kt#L47-L80

**Comment by kevinhaube** [comment]
@chris-kuryak @stephenkao
Our historical holdup on this is as follows:

1. Drew's concern of a lack of larger onboarding automation/self-servicing planning
> This concern, to me, is non-blocking. We have a tool, we should make it work to suit the requirements and release it so O&O may direct users to it.
2. No feedback on the initial development: we are unaware if what we have helps or not

In theory, yes, it'd be a huge gain for onboarding! What we have now, though, is limited:

- **ONLY** validates against `[org].default` sender (`default` is the sender setting name)
- **REQUIRES** a user be logged in because it's using `MembershipSettings` to pull `orgname` to even _get_ the `default` sender setting

Neither of these seem in any way helpful to somebody logged out. Ideally what we'd want to know is:

- WHAT do they validate against? 
- Do you not already have a setting set up to validate? (As stated above, currently, we **are** validating against already-established settings, currently.)

**Comment by chris-kuryak** [comment]
@GilmoreA6 @brick-green : Check out the technical questions from @kevinhaube in the previous comment:

> Ideally what we'd want to know is:

1. Who's settings do they validate a file against?
2. Do you already have a setting set up to validate? (As stated above, currently, we are validating against already-established settings.)
3. Is it okay for us to expose org settings, perhaps to read what available settings a user can validate against, via an unauthenticated page?

Edit: I believe we spoke of a "generic" schema we'd make public, something like validation-service.hl7-standard or something akin to it. This would be a great option if we come to the conclusion that a unique setting to validate against isn't required for each prospective sender. >

Do y'all have answers on these? Or is there a better forum to ask these questions?

**Comment by GilmoreA6** [comment]
@kevinhaube 
1.) We currently have three "master schemas" 1 for HL7, 1 for CSV and 1 for CSV OTC. We are attempting to keep it to those three schemas where possible. We currently validate against one of those three or we create a custom schema if necessary (again we like to avoid doing this)
2.) When validating initially we either set up the org locally with one of the three schemas above (or custom schema if we already know they will need one) or we use the CLI tools to validate against the expected schema.
3.) Personally, if we could just have a way to switch between the three "master" schema then the user could pick one of the three that would be great. If they need a custom they would have to contact us anyway so we could handle those on a case by case basis. (The three master schemas would be HL7/hl7-ingest-covid-19-training.schema, upload-covid-19.schema, csv-otc-covid-19.schema)

**Comment by kevinhaube** [comment]
@GilmoreA6 Thank you so much for the info! This is super helpful for technical planning.

@stephenkao @chris-kuryak Happy to assist turning the info into actionable engineering tickets once I'm back 🙂 Looks like maybe it'll pan out like:

- Create a "public" organization that has three senders: one per standard schema. (Can be done via UI, thanks to Tom N's Org settings interfaces.)
- In the Validator component system, we should fetch this public org settings entry and it's three sender settings entries to be swapped between

By this point, public validation cases should be covered. For the remaining nuanced case of needing a particular schema, **with the assumption that Operations has made them an org and sender settings entry**:

(Requires login)

- Create **either** a generic group like DHvalidating **or** their own orgs group per our standard for web app users, DHmy-org, and add the user who needs to validate a special schema. 
- Create a system that can retrieve the custom settings entries for validation

The reasoning for putting custom work behind login is that we do not want a long list of public schemas as it sends the wrong message about our standards.

For okta groups naming: I'd recommend just the DHvalidating group as we do not want to create and destroy groups repeatedly. That may lead to many stale and empty groups to clean up later. With this model in mind, the settings would be stored as follows:

- public org: 3 senders, standard schemas
- validating org: n-many senders as there are custom schemas being validated against _at the moment_. 

The public org would need to be retrieved without a bearer token or admin auth; this part may be on the Platform team to give us a bypass for "public" org name. 

Validating org would still be behind auth as there would be an Okta group tied to it for login and auth purposes.

**Comment by chris-kuryak** [comment]
Removing as blocked per answer from James and syncing with Exp engr team.

**Comment by chris-kuryak** [comment]
Per scoping session with @kevinhaube @stephenkao, we'll be breaking this into a small epic with several smaller engineering and design issues.
